packages/lib/gpcircuits/circuits/list-membership.circom

```diff
@@ -14,25 +16,51 @@ template ListMembershipModule(
     // Value to be checked.
     signal input comparisonValue; 
 
-    // List of admissible value hashes. Assumed to have repetitions if the actual list length is smaller.
+    // List of admissible value hashes. Assumed to have repetitions if
+    // the actual list length is smaller.
     signal input validValues[MAX_LIST_ELEMENTS]; 
 
-    // Boolean indicating whether `comparisonValue` lies in `validValues`.
+    // Boolean indicating whether `comparisonValue` lies in
+    // `validValues`.
     signal output isMember;
 
-    signal partialProduct[MAX_LIST_ELEMENTS];
+    isMember <== IsZero()(
+        UnnormalisedListNonMembership(MAX_LIST_ELEMENTS, MAX_LIST_ELEMENTS)(
+            comparisonValue,
+            validValues
+        )
+    );
+}
 
-    for (var i = 0; i < MAX_LIST_ELEMENTS; i++) {
-        if (i == 0) {
-            partialProduct[i] <== comparisonValue - validValues[i];
-        } else {
-            partialProduct[i] <== partialProduct[i-1] * (comparisonValue - validValues[i]);
-        }
-    }
+/**
+ * Helper template returning a non-zero field element if the given
+ * value is not an element of the given list restricted to the first
+ * `NUM_LIST_ELEMENTS` elements and zero otherwise. This is done by
+ * subtracting the value from all elements of the list and folding it
+ * by means of field multiplication.
+ */
+template UnnormalisedListNonMembership(NUM_LIST_ELEMENTS, MAX_LIST_ELEMENTS) {
+     // Value to be checked.
+    signal input comparisonValue; 
+
+    // List of admissible values.
+    signal input validValues[MAX_LIST_ELEMENTS];
+
+    // Indicator of whether the value is not an element of the list of
+    // admissible values, viz. a non-zero field element iff the value
+    // is a non-member.
+    signal output isNotMember;
 
-    if (MAX_LIST_ELEMENTS == 0) {
-        isMember <== 0;
+    if (NUM_LIST_ELEMENTS == 0) {
+        isNotMember <== 1;
     } else {
-        isMember <== IsZero()(partialProduct[MAX_LIST_ELEMENTS - 1]);
+        isNotMember <== MultiAND(NUM_LIST_ELEMENTS)(
+            Add(NUM_LIST_ELEMENTS)(
+                -comparisonValue,
+                Take(NUM_LIST_ELEMENTS, MAX_LIST_ELEMENTS)(
```

Comment:
This is how I initially did it, but Circom doesn't like that approach:
```
thread 'main' panicked at compiler/src/intermediate_representation/translate.rs:959:25:
internal error: entered unreachable code: On development: Circom compiler does not accept for now the assignment of arrays of unknown sizes during the execution of loops
```

Reply:
Where's the unknown size?  I'd expect all the array lenghts to be known (based on template args) at the point of use.  I'd have to see the specific code to comment on it, though, and I'm fully willing to believe that circom is broken.

Reply:
All of the array sizes depend only on the template parameter and loop index. I assume there is a bug in the loop + anonymous component unrolling code. What I did was replace the list non-membership and uniqueness module templates with the following:

```circom
template ListMembershipModule(
    MAX_LIST_ELEMENTS
) {
    signal input comparisonValue; 

    signal input validValues[MAX_LIST_ELEMENTS]; 

    signal output isMember;

    isMember <== IsZero()(
        NotEqualsAny(MAX_LIST_ELEMENTS)(
            comparisonValue,
            validValues
        )
    );
}

template UniquenessModule(
    NUM_LIST_ELEMENTS
) {
    signal input values[NUM_LIST_ELEMENTS];

    signal output valuesAreUnique;

    signal noDupsAfter[NUM_LIST_ELEMENTS];

    for(var i = 0; i < NUM_LIST_ELEMENTS; i++) {
        var j = i+1;
        noDupsAfter[i] <==
            NotEqualsAny(NUM_LIST_ELEMENTS - j)(
                values[i],
                Take(NUM_LIST_ELEMENTS - j, NUM_LIST_ELEMENTS)(
                    ArrayRotl(j, NUM_LIST_ELEMENTS)(values)
                )
            );
    }

    valuesAreUnique <== NOT()(
        IsZero()(
            MultiAND(NUM_LIST_ELEMENTS)(
                noDupsAfter
            )
        )
    );
}
```
My original version used `Drop` instead of `Take` (which is what I believe you suggested in your other comment) but yielded the same result.

Reply:
Interesting.  Seems like a circom limitation keeps us from sizing arrays based on loop variables, but not if they're inside of templates.  It's unintuitive to me that a loop variable can be a template argument (as demonstrated in our use of Take and Rotl) but can't be an array size.  Take does return an array of a size which depends on its template argument, so it shouldn't matter if it's directly in the loop or in a template instantiated in the loop.

Would it help if you combined Take and Rotl into a single Slice template which returned array elements from [I, I+N)?  Or would that make the situation worse?  I don't really understand what circom's limitation is, so it's hard to know what to suggest.  If there's no obvious solution I'm okay with accepting what works, given the final number of constraints is the same.

Reply:
I tried doing this with a single [`drop`](https://srfi.schemers.org/srfi-1/srfi-1.html#drop)-like template but ran into the same error, which is what led me to the current approach.